<a href="https://colab.research.google.com/github/shopy100/machineLearning/blob/main/Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Pandas Tutorial



In [ ]:
#Load a CSV file into a Pnadas DataFrame:
import pandas as pd
df = pd.read_csv('data.csv')
print(df.to_string())

           Car       Model  Volume  Weight  CO2
0       Toyoty        Aygo    1000     790   99
1   Mitsubishi  Space Star    1200    1160   95
2        Skoda      Citigo    1000     929   95
3         Fiat         500     900     865   90
4         Mini      Cooper    1500    1140  105
5           VW         Up!    1000     929  105
6        Skoda       Fabia    1400    1109   90
7     Mercedes     A-Class    1500    1365   92
8         Ford      Fiesta    1500    1112   98
9         Audi          A1    1600    1150   99
10     Hyundai         I20    1100     980   99
11      Suzuki       Swift    1300     990  101
12        Ford      Fiesta    1000    1112   99
13       Honda       Civic    1600    1252   94
14      Hundai         I30    1600    1326   97
15        Opel       Astra    1600    1330   97
16         BMW           1    1600    1365   99
17       Mazda           3    2200    1280  104
18       Skoda       Rapid    1600    1119  104
19        Ford       Focus    2000    13

Pandas is a Python library used fro working with data sets.




It has fucntions for analyzing, cleaning, exploring, and manipulating data.

The name "Pandas" has a reference to both "Panel Data", and "Python Data Analysis" and we creaetd by Wes Mckinney in 2008.



---

#WHY USE PANDAS?
Pandas allows us to analyze big data and make conclusions based on statistical theroies.

Pandas can clean messy data sets, and make them readable and relevant.

Relevant data is very important in data science.

> **DATA SCIENCE** : is a branch of computer science where we study how to store, use and analyze data for deriving information from it.



---

#WHAT CAN PANDAS DO?

Pandas gives you answers about the data. Like:
- Is there a correlation between two or more columns?
- What is average value?
- Max value?
- Min value?

Pandas are also able to delete that are not relevant, or contains wrong values, like empty or NULL values.
This is called cleaning the data.


---

Where is the Pandas Codebase?
[github](https://github.com/shopy100/pandas)

> github: enables many people to work on the same codebase.

#IMPORT PANDAS


In [ ]:
import pandas

In [ ]:
import pandas
mydataset = {
  'cars': ["BMW", "Volvo", "Ford"],
  'passings': [3, 7, 2]
}

myvar = pandas.DataFrame(mydataset)

print(myvar)

    cars  passings
0    BMW         3
1  Volvo         7
2   Ford         2


#Pandas as pd

Pandas is usually imported under the pd alias.

> **alias** : in Python alias are an ultimate name for referring to the smae thing.

Create an alias with the as keyword while importing:

```import pandas as pd```

Now the Pandas package can be referred to as pd instead of pandas

In [ ]:
import pandas as pd

mydataset = {
    'cars': ["BMW", "Volvo", "Ford"],
  'passings': [3, 7, 2]
}

myvar = pd.DataFrame(mydataset)

print(myvar)


    cars  passings
0    BMW         3
1  Volvo         7
2   Ford         2


#CHECKING PANDAS VERSION

The version string is stored under __version__ attribute

In [ ]:
import pandas as pd
print(pd.__version__)

2.2.3


#**PANDAS SERIES**

A pandas Series is like a column in a table.

It is a one-diemnsional array holding data of any type.

In [ ]:
#Create a simple Pandas Series from a list

import pandas as pd

a = [1, 7, 2]

myvar = pd.Series(a)

print(myvar)

0    1
1    7
2    2
dtype: int64


#**LABELS**

If nothing else is specified, the values are labeled with their index number.

First value gas index 0, second value has index 1 etc.

This label can be used to access a specified value.

In [ ]:
#Return the first value of teh Series
import pandas as pd
a = [1, 7, 2]

myvar = pd.Series(a)

print(myvar[0])

1


#**CREATE LABELS**
With the ```argument```, you can name your own labels.

In [ ]:
#Create your own labels:
import pandas as pd

a = [1, 7, 2]

myvar = pd.Series(a, index=["x", "y", "z"])

print(myvar)

x    1
y    7
z    2
dtype: int64


**when you have created labels, you can access an item by referring to the label**

In [ ]:
#Return the valid of "y"
import pandas as pd

a = [1, 7, 2]

myvar = pd.Series(a, index=["x", "y", "z"])

print(myvar["y"])

7


#**KEY/VALUE OBJECTS AS SERIES**

You can also use a key/value object, like a dictionary, when creating a Series.

In [ ]:
#Create a simple Pandas Series from a dictionary:
import pandas as pd

calories = {"day1": 420, "day2": 380, "day3": 390}

myvar = pd.Series(calories)

print(myvar)

day1    420
day2    380
day3    390
dtype: int64


**Note** : The keys of the dictionary become the labels



**To select only some of the items in the dictionary, use the ```index``` argument and specify only the items you want to include in the Series.**

In [ ]:
#Create a Series using only data from "day1" and "day2":
import pandas as pd

calories = {"day1": 420, "day2": 380, "day3": 390}

myvar = pd.Series(calories, index = ["day1", "day2"])

print(myvar)

420


/tmp/ipykernel_2406/2804500724.py:8: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  print(myvar[0])


#**DataFrames**

Data sets in Pandas are usually multi-dimensional tabels, called DataFrames.

Series is like a column, a DataFrame is the whole table.

In [ ]:
#Create a DataFrame from two Series:
import pandas as pd

data = {
    "calories": [420, 380, 390],
    "duration": [50, 40, 45]
}

myvar = pd.DataFrame(data)

print(myvar)


   calories  duration
0       420        50
1       380        40
2       390        45


#**DataFrames**

A Pandas DataFrame is a 2 dimensional data structure, like a dimensional array, or a table with rows and columns.

In [ ]:
#Create a simple Pandas DataFrame:
import pandas as pd
data = {
    "calories": [420, 380, 390],
    "duration": [50, 40, 45]
}

#load data into a DataFrame object:
df = pd.DataFrame(data)

print(df)


   calories  duration
0       420        50
1       380        40
2       390        45


#**LOCATE ROW**

As you can see from the reuslt above, the DataFrame is like a table with rows and columns.

Pandas use the ```loc``` attribute to return one or more specified row(s)

In [ ]:
import pandas as pd
data = {
    "calories": [420, 380, 390],
    "duration": [50, 40, 45]
}

#load data into a DataFrame object:
df = pd.DataFrame(data)

#refer to the row index
print(df.loc[0])

calories    420
duration     50
Name: 0, dtype: int64


**Note** : This example returns a Pandas **Series**

In [ ]:
#Return row 0 and 1:
import pandas as pd
data = {
    "calories": [420, 380, 390],
    "duration": [50, 40, 45]
}
df = pd.DataFrame(data)

#use a list of indexes
print(df.loc[[0, 1]])

   calories  duration
0       420        50
1       380        40


Note: When using [], the result is a Pandas **DataFrama**

#**NAMED INDEXES**

With the ```index``` argument, you can name your own indexes

In [ ]:
#Add a list if names to give each row a name
import pandas as pd
data = {
    "calories": [420, 380, 390],
    "duration": [50, 40, 45]
}
df = pd.DataFrame(data, index = ["day1", "day2", "day3"])

print(df)

      calories  duration
day1       420        50
day2       380        40
day3       390        45


#**LOCATE NAMED INDEXES**

Use the names index in the ```loc``` attribute to return the specified row(s)

In [ ]:
#Return "day2":
import pandas as pd
data = {
    "calories": [420, 380, 390],
    "duration": [50, 40, 45]
}
df = pd.DataFrame(data, index = ["day1", "day2", "day3"])

print(df.loc["day2"])

calories    380
duration     40
Name: day2, dtype: int64


#**LOAD FILES INTO A DataFrame**

If your data sets are stored in a file, Pandas can load them into a DataFrame

In [ ]:
#Load a comma separated file (CSV file) into a DataFrame:
import pandas as pd

df = pd.read_csv('data.csv')

print(df)

           Car       Model  Volume  Weight  CO2
0       Toyoty        Aygo    1000     790   99
1   Mitsubishi  Space Star    1200    1160   95
2        Skoda      Citigo    1000     929   95
3         Fiat         500     900     865   90
4         Mini      Cooper    1500    1140  105
5           VW         Up!    1000     929  105
6        Skoda       Fabia    1400    1109   90
7     Mercedes     A-Class    1500    1365   92
8         Ford      Fiesta    1500    1112   98
9         Audi          A1    1600    1150   99
10     Hyundai         I20    1100     980   99
11      Suzuki       Swift    1300     990  101
12        Ford      Fiesta    1000    1112   99
13       Honda       Civic    1600    1252   94
14      Hundai         I30    1600    1326   97
15        Opel       Astra    1600    1330   97
16         BMW           1    1600    1365   99
17       Mazda           3    2200    1280  104
18       Skoda       Rapid    1600    1119  104
19        Ford       Focus    2000    13

#**READ CSV**

##READ CSV FILES

- A simple way to store big data sets is to use CSV files (comma spearated files).
- CSV files contains plain text and is a well know format that can be read by everyone including Pandas.


In [ ]:
#Load the CSV into a DataFrame
import pandas as pd
df = pd.read_csv('data.csv')
print(df.to_string())

           Car       Model  Volume  Weight  CO2
0       Toyoty        Aygo    1000     790   99
1   Mitsubishi  Space Star    1200    1160   95
2        Skoda      Citigo    1000     929   95
3         Fiat         500     900     865   90
4         Mini      Cooper    1500    1140  105
5           VW         Up!    1000     929  105
6        Skoda       Fabia    1400    1109   90
7     Mercedes     A-Class    1500    1365   92
8         Ford      Fiesta    1500    1112   98
9         Audi          A1    1600    1150   99
10     Hyundai         I20    1100     980   99
11      Suzuki       Swift    1300     990  101
12        Ford      Fiesta    1000    1112   99
13       Honda       Civic    1600    1252   94
14      Hundai         I30    1600    1326   97
15        Opel       Astra    1600    1330   97
16         BMW           1    1600    1365   99
17       Mazda           3    2200    1280  104
18       Skoda       Rapid    1600    1119  104
19        Ford       Focus    2000    13

**Tip**: use ```to_string()``` to print the entire DataFrame

If you have a large DataFrame with many rows, Pandas will only retunr the first the row 5 rows, and the last 5 rows.

In [ ]:
#Print the DataFrame without the ```to_string()``` method:
import pandas as pd
df = pd.read_csv('data.csv')
print(df)

           Car       Model  Volume  Weight  CO2
0       Toyoty        Aygo    1000     790   99
1   Mitsubishi  Space Star    1200    1160   95
2        Skoda      Citigo    1000     929   95
3         Fiat         500     900     865   90
4         Mini      Cooper    1500    1140  105
5           VW         Up!    1000     929  105
6        Skoda       Fabia    1400    1109   90
7     Mercedes     A-Class    1500    1365   92
8         Ford      Fiesta    1500    1112   98
9         Audi          A1    1600    1150   99
10     Hyundai         I20    1100     980   99
11      Suzuki       Swift    1300     990  101
12        Ford      Fiesta    1000    1112   99
13       Honda       Civic    1600    1252   94
14      Hundai         I30    1600    1326   97
15        Opel       Astra    1600    1330   97
16         BMW           1    1600    1365   99
17       Mazda           3    2200    1280  104
18       Skoda       Rapid    1600    1119  104
19        Ford       Focus    2000    13

#**max_rows**

- The number of rows returned is defined in Pandas option settings.
- You can check your system's maximum rows with the ```pd.option.display.max_rows``` statement

In [ ]:
#Check the number of maximum returned rows:
import pandas as pd
print(pd.options.display.max_rows)

60


- In my system the number is 60, which means that if the DataFrame contains more than 60 rows, the ```print(df)``` statement will return only the headers and the first and last 5 rows.
- You can change the maximum rows number with the same statement.


In [ ]:
#Increase the maximum number of rows to display the entire DataFrame:
import pandas as pd
pd.options.display.max_rows = 9999
df = pd.read_csv('data.csv')
print(df)

           Car       Model  Volume  Weight  CO2
0       Toyoty        Aygo    1000     790   99
1   Mitsubishi  Space Star    1200    1160   95
2        Skoda      Citigo    1000     929   95
3         Fiat         500     900     865   90
4         Mini      Cooper    1500    1140  105
5           VW         Up!    1000     929  105
6        Skoda       Fabia    1400    1109   90
7     Mercedes     A-Class    1500    1365   92
8         Ford      Fiesta    1500    1112   98
9         Audi          A1    1600    1150   99
10     Hyundai         I20    1100     980   99
11      Suzuki       Swift    1300     990  101
12        Ford      Fiesta    1000    1112   99
13       Honda       Civic    1600    1252   94
14      Hundai         I30    1600    1326   97
15        Opel       Astra    1600    1330   97
16         BMW           1    1600    1365   99
17       Mazda           3    2200    1280  104
18       Skoda       Rapid    1600    1119  104
19        Ford       Focus    2000    13

#**READ JSON**

##READ JSON

- Big data sets are often stored, or extracted as JSON.
- JSON is plain text, but has the format of an object, and is well known in the world of programming, including Pandas.


In [ ]:
#Load the JSON file into a DataFrame:
import pandas as pd
df = pd.read_json('example.json')
print(df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.5
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

**Tip** : use to_string() to print the entire DataFrame.

##**DICTIONARY as JSON**

**JSON = Python Dictionary**

JSON objects have the same format as Python dictioanries.

if your JSON code is not in a file, but in a Python Dictionary, you can load it into a DataFrame directly:

In [ ]:
#Load a Python Dictionary into a DataFrame:
import pandas as pd

data = {
    "Duration":{
        "0":60,
        "1":60,
        "2":60,
        "3":45,
        "4":45,
        "5":60
    },
    "Pulse":{
        "0":110,
        "1":117,
        "2":103,
        "3":109,
        "4":117,
        "5":102
    },
    "Maxpulse":{
        "0":130,
        "1":145,
        "2":135,
        "3":175,
        "4":148,
        "5":127
    },
    "Calories":{
        "0":409,
        "1":479,
        "2":340,
        "3":282,
        "4":406,
        "5":300
    }
}

df = pd.DataFrame(data)

print(df)

   Duration  Pulse  Maxpulse  Calories
0        60    110       130       409
1        60    117       145       479
2        60    103       135       340
3        45    109       175       282
4        45    117       148       406
5        60    102       127       300


#**ANALYZING DataFrames**
##**VIEWING THE DATA**

- One of the most used method for getting a wuick overview of the DataFrame, is the ```head()``` method.
- The ```head()``` method returns the headers and a specified number of rows, starting from the top.

In [ ]:
#Get a quick overview by printing the first 10 rows of the DatafRAME:
import pandas as pd
df = pd.read_csv('data.csv')
print(df.head(10))

          Car       Model  Volume  Weight  CO2
0      Toyoty        Aygo    1000     790   99
1  Mitsubishi  Space Star    1200    1160   95
2       Skoda      Citigo    1000     929   95
3        Fiat         500     900     865   90
4        Mini      Cooper    1500    1140  105
5          VW         Up!    1000     929  105
6       Skoda       Fabia    1400    1109   90
7    Mercedes     A-Class    1500    1365   92
8        Ford      Fiesta    1500    1112   98
9        Audi          A1    1600    1150   99


**Note** : if the number of rows is not specified, the ```head()``` method will return the top 5 rows.

In [ ]:
#Print the first 5 rows of the DataFrame:
import pandas as pd
df = pd.read_csv('data.csv')
print(df.head())

          Car       Model  Volume  Weight  CO2
0      Toyoty        Aygo    1000     790   99
1  Mitsubishi  Space Star    1200    1160   95
2       Skoda      Citigo    1000     929   95
3        Fiat         500     900     865   90
4        Mini      Cooper    1500    1140  105


- there is also a ```tail()``` method for viewing the *last* rows of the DataFrame
- the ```tail()``` method returns the headers and a specified number of rows, starting from the bottom

In [ ]:
#Print the last 5 rows of the DataFrame:
import pandas as pd
df = pd.read_csv('data.csv')
print(df.tail())

         Car   Model  Volume  Weight  CO2
31     Volvo    XC70    2000    1746  117
32      Ford   B-Max    1600    1235  104
33       BMW     216    1600    1390  108
34      Opel  Zafira    1600    1405  109
35  Mercedes     SLK    2500    1395  120


#**INFO ABOUT THE DATA**

The DataFranes object has a method called ```info()```, that gives you mmore information about the data set.


In [ ]:
#Print infomration about the data:
import pandas as pd
df = pd.read_csv('data2.csv')
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 169 entries, 0 to 168
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  169 non-null    int64  
 1   Pulse     169 non-null    int64  
 2   Maxpulse  169 non-null    int64  
 3   Calories  164 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 5.4 KB
None


#**RESULT EXPLAINED**

- The result tells us thee are 169 rows and 4 columns:



> RangeIndex: 169 entries, 0 to 168
Data columns (total 4 columns):

And the name of each coumn, with the data type:
#   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  169 non-null    int64  
 1   Pulse     169 non-null    int64  
 2   Maxpulse  169 non-null    int64  
 3   Calories  164 non-null    float64






#**NULL VALUES**

- The ```info()``` method also tells us how many Non-Null values there are present in each column, and in our data set it seems like there are 164 of 169 Non-Null values in the 'Calories' column
- Which means that there are 5 rows with no value at all, in the "Calories" column, for whatever reason.
- Empty values, or Null values, can be bad when analyzing data, and you should consider removing rows with empty values.
> This is a step towards what is called *cleaning data*, and you will learn more about that in the next chapters.

#**CLEANING DATA**

##DATA CLEANING

- Data cleaning means fixing bad daa in your data set.
- Bad data could be:
> - Empty celss
> - Data in wrong format
> - Wrong data
> - Duplicates

- In this tutorial you will learn how to deal with all of them.



---
##OUR DATA SET

will use data2.csv

- The data set contains some empty celss ("Date" in row 22, and "Calories" in row 18 and 28).
- The data set contains wrong fromat ( "Date" in row 26)
- The data set contains wrong data ("Duration" in row 7)
- The data set contains duplicates(row 11 and 12)


#** CLEANING EMPTY CELLS**
##EMPTY CELLS
Empty cells can potentially give you a wrong result when you analyze data
---
##REMOVE ROWS
- one way to deal with empty cells is to remove rows that contain empty cells
-  This is usually OK, since data sets can be very big, and removing a few rows will not have a big impact on the result.


In [ ]:
#Return a new Data Frame with no empty cells:
import pandas as pd
df = pd.read_csv('data2.csv')
new_df = df.dropna()
print(new_df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45    105       132     246.0
24         60   

**Note** : By default, the ```dropna()``` method returns a *new* DataFrame, and will not change the original

If you want to change the original DataFrame, use the ```inplace=True``` argument

In [ ]:
#Remove all wors with NULL values:
import pandas as pd
df = pd.read_csv('data2.csv')
df.dropna(inplace = True)
print(df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45    105       132     246.0
24         60   

**Note** : Now, the ```dropna(inplace = True)``` will NOT return a new DataFrame, but it will remove all rows containing NULL values from the original DataFrame.



#**REPLACE EMPTY VALUES**

- Another way of dealing with empty cells is to insert a *new* value instead.
- This way you do not have to delete entire rows just because of some empty cells.

- The ```fillna()``` method allows us to replace empty cells with a value:

In [ ]:
#Replace NULL values with the number 130:
import pandas as pd
df = pd.read_csv('data2.csv')
df.fillna(130, inplace = True)
print(df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112     130.0
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

#**REPALCE ONLY FOR SPECIFIED COLUMNS**
- The example above replaces all empty cells in the whole Data Frame
- To only replace empty values for one column, specify the *column* name for the DataFrame:


In [ ]:
#Replace nULL values in the "Calories" columns with the number 130:
import pandas as pd
df = pd.read_csv('data2.csv')
df.fillna({"Calories": 130}, inplace=True)
print(df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112     130.0
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

#**REPLACE USING MEAN, MEDIAN, OR MODE
- A Common way to replace empty cells, is to calculate the mean, median or mode value of the column
- Pandas uses the ```mean(), median() and mode()``` methods to calculate the respective values for a specified column:

In [ ]:
#Calculate the MEAN, and replace any empty values with it:
import pandas as pd
df = pd.read_csv('data2.csv')
x = df["Calories"].mean()
df.fillna({"Calories": x}, inplace=True)
print(df.to_string())

     Duration  Pulse  Maxpulse     Calories
0          60    110       130   409.100000
1          60    117       145   479.000000
2          60    103       135   340.000000
3          45    109       175   282.400000
4          45    117       148   406.000000
5          60    102       127   300.000000
6          60    110       136   374.000000
7          45    104       134   253.300000
8          30    109       133   195.100000
9          60     98       124   269.000000
10         60    103       147   329.300000
11         60    100       120   250.700000
12         60    106       128   345.300000
13         60    104       132   379.300000
14         60     98       123   275.000000
15         60     98       120   215.200000
16         60    100       120   300.000000
17         45     90       112   375.790244
18         60    103       123   323.000000
19         45     97       125   243.000000
20         60    108       131   364.200000
21         45    100       119  

**Mean** = the average value (the sum of all values divided by number of values).

In [ ]:
#Calculate the MEDIAN, and repalce any empty values with it:
import pandas as pd
df = pd.read_csv('data2.csv')
x = df["Calories"].median()
df.fillna({"Calories": x}, inplace=True)
print(df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112     318.6
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

**Median** = the value in the middle, after you have sorted all values ascending

In [ ]:
#Calculate the MODE, and repalce any empty values with it:
import pandas as pd
df = pd.read_csv('data2.csv')
x = df["Calories"].mode()[0]
df.fillna({"Calories": x}, inplace=True)
print(df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112     300.0
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

**Mode** = the value that appears most frequently

#**CLEANING DATA OF WRONG FORMAT**

###DATA OF WRONG FORMAT

- Cells with data of wrong can make it difficult, or even impossible, to analyze data.
- To fix it, you have two options:
> - remove the rows,
> - convert all cells in the columns into the same format

---
#CONVERT INTO A CORRECT FORMAT
- In our Data Frame, we have two cells with the wrong format.
- Check out row 22 and 26, the 'Date' column should be a string that represents a date:




In [ ]:
import pandas as pd
df = pd.read_csv('data2_with_date.csv')
print(df.to_string())

     Duration  Pulse  Maxpulse  Calories                Date;
0          60    110       130     409.1          2020/08/04;
1          60    117       145     479.0          2025/02/22;
2          60    103       135     340.0          2024/05/08;
3          45    109       175     282.4          2022/12/02;
4          45    117       148     406.0          2022/11/17;
5          60    102       127     300.0          2025/09/16;
6          60    110       136     374.0          2020/07/27;
7          45    104       134     253.3          2024/08/20;
8          30    109       133     195.1       2021/05/04;Nan
9          60     98       124     269.0          2020/08/16;
10         60    103       147     329.3          2023/07/02;
11         60    100       120     250.7          2026/06/27;
12         60    106       128     345.3          2024/11/22;
13         60    104       132     379.3          2025/01/22;
14         60     98       123     275.0       2024/10/08;NaN
15      

- Let;s try to convert all cells in the 'Date' column into dates

- Pandas has a ```to_datetime()``` method for this:

In [ ]:
#Convert to date:
import pandas as pd
df = pd.read_csv('data2_with_date.csv')
df['Date'] = pd.to_datetime(df['Date'], format='mixed')
print(df.to_string())

DateParseError: Unknown datetime string format, unable to parse: 2021/05/04;Nan, at position 8

- As you can see from the result, the date in row 26 was mixed, but the empty date in row 22 got a NaT (Not a Time) value, in other words an empty value.
- One way to deal with empty values is simply removing the entire row.


#**REMOVING ROWS**

- the result from the converting in the example above gave us a NaT value, which can be handled as a NULL value, and we can remove the row by using the ```dropna()``` method.

In [ ]:
#Remove rows with a NULL value in the "Date" column:
df.dropna(subset=['Date'], inplace = True)

print(df.to_string())

KeyError: ['Date']

#**FIXING WRONG DATA**
#WRONG DATA
- "Wrong data" does not have to be "empty cells" or "wrong format", it can just be worng, like if someone registered "199" instead of "1.99".
- sometimes you can spot worng data by looking at the data set, because you have an expection of what it should be.
- if you take a look at our data set, you can see that in row 7, the duration is 450, but for all the other rows the duration is between 30 and 60.
- it doesn't have to be wring, but taking in consideration that this is data set of seomeone's workout sessions, we conclude with the fact that this person did not work out in 450 minutes.
---
- how we can we fix wrong values, like the one for "Duration" in row 7?
---
#**Replacing values**
- One way to fix wrong values is to replace them with something else.
- In our example, it is most likely a typo, and the value should be "45" instead of "450", and we could insert "45" in row 7:
```
Set"Duration" = 45 in row 7:
df.loc[7, 'Duration'] = 45
```
---
- For small data sets you wight be able to rpelace the worng dat aone by one, but not for big data sets.
- To replace wrong data for larger data sets you can create some rules, e.g. set some boundaries for legal values, and replace any values that are outside of the boundaries.
---
- Loop through all values in the "Duration" column.
- If the value is higher than 120, set it to 120:
```
for x in df.index:
  if df.loc[x, "Duration"] > 120:
     df.loc[x, "Duration"] = 120
```

In [2]:
import pandas as pd
df = pd.read_csv('data2.csv')
for x in df.index:
  if df.loc[x, "Duration"] > 120:
    df.loc[x, "Duration"] = 120

print(df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       120     409.1
1          60    117       120     479.0
2          60    103       120     340.0
3          45    109       120     282.4
4          45    117       120     406.0
5          60    102       120     300.0
6          60    110       120     374.0
7          45    104       120     253.3
8          30    109       120     195.1
9          60     98       120     269.0
10         60    103       120     329.3
11         60    100       120     250.7
12         60    106       120     345.3
13         60    104       120     379.3
14         60     98       120     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       120     323.0
19         45     97       120     243.0
20         60    108       120     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

###**REMOVING ROWS**
- Another way of handling wrong data is to remove that contains wrong data.
- This way yo do not have to find out what to replace them with, and there is a good chance you do not need them to do your analyses.

In [3]:
#Delete rows where "Duration" is higher than 120:
import pandas as pd
df = pd.read_csv('data2.csv')

for x in df.index:
  if df.loc[x, "Duration"] > 120:
    df.drop(x, inplace = True)

print(df.to_string())


     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

#**REMOVING DUPLICATES**